src/Policies.cpp

```diff
@@ -96,17 +96,79 @@ std::vector<Policy *> PolicyManager::load(const YAML::Node &policy_yaml)
             throw PolicyException("missing or invalid policy input stream configuration at key 'input'");
         }
         auto input_node = it->second["input"];
-        if (!input_node["tap"] || !input_node["tap"].IsScalar()) {
-            throw PolicyException("missing or invalid tap at key 'input.tap'");
-        }
         if (!input_node["input_type"] || !input_node["input_type"].IsScalar()) {
             throw PolicyException("missing or invalid input_type at key 'input.input_type'");
         }
 
-        // Tap
-        Tap *tap{nullptr};
-        std::unique_lock<std::shared_mutex> tap_lock;
-        auto tap_name = input_node["tap"].as<std::string>();
+        if (input_node["tap"] && input_node["tap_selector"]) {
+            throw PolicyException("input can have only key 'input.tap' or key 'input.tap_selector'");
+        } else if (!input_node["tap"] && !input_node["tap_selector"]) {
+            throw PolicyException("missing key 'input.tap' or key 'input.tap_selector'");
+        }
+
+        if (input_node["tap"]) {
+            if (!input_node["tap"].IsScalar()) {
+                throw PolicyException("invalid tap at key 'input.tap'");
+            }
+            result.push_back(create_policy(it->second, policy_name));
+        } else if (input_node["tap_selector"]) {
+            auto tap_selector = input_node["tap_selector"];
+            if (!tap_selector.IsMap()) {
+                throw PolicyException("'input.tap_selector' is not a map");
+            }
+
+            if (tap_selector["all"] && tap_selector["any"]) {
+                throw PolicyException("input can have only key 'input.tap_selector.all' or key 'input.tap_selector.any'");
+            } else if (!tap_selector["all"] && !tap_selector["any"]) {
+                throw PolicyException("missing key 'input.tap_selector.all' or key 'input.tap_selector.any'");
+            }
+
+            bool all = true;
+            std::string binary_op;
+            if (tap_selector["all"]) {
+                if (!tap_selector["all"].IsMap()) {
+                    throw PolicyException("'input.tap_selector.all' is not a map");
+                }
+                all = true;
+                binary_op = "all";
+            } else if (tap_selector["any"] || tap_selector["any"].IsMap()) {
+                if (!tap_selector["any"].IsMap()) {
+                    throw PolicyException("'input.tap_selector.any' is not a map");
+                }
+                all = false;
+                binary_op = "any";
+            }
+
+            bool match = false;
+            auto [tap_modules, hm_lock] = _registry->tap_manager()->module_get_all_locked();
+            for (auto &[name, mod] : tap_modules) {
+                auto tmod = dynamic_cast<Tap *>(mod.get());
+                if (tmod && tmod->tags_validate_yaml(tap_selector[binary_op], all)) {
+                    auto policy_selector_name = policy_name + "_" + tmod->name();
```

Comment:
I think we are going to have to find a way to make this policy name stay the same, but have it have an additional label for the tap it used. if we don't do this, it will be poor UX experience in the dashboards, with an ever expanding list of these policy-tap prefixes in the Policy dropdown

I believe some of our other use cases are also forcing us to think about the label issue, so hopefully there is a common solution

Reply:
In any case, we should also think about JSON return, how labels will work for it

Reply:
done